c_src/emlx_nif.cpp

```diff
@@ -479,6 +507,73 @@ NIF(eval) {
   return nx::nif::ok(env);
 }
 
+NIF(set_compile) {
+  PARAM(0, bool, compile);
+
+  if (compile) {
+    mlx::core::enable_compile();
+  } else {
+    mlx::core::disable_compile();
+  }
+
+  return nx::nif::ok(env);
+}
+
+void move_between_envs(ERL_NIF_TERM from_term, ERL_NIF_TERM *to_term,
+                       ErlNifEnv *from_env, ErlNifEnv *to_env) {
+  ErlNifBinary serialized;
+  enif_term_to_binary(from_env, from_term, &serialized);
+  enif_binary_to_term(to_env, serialized.data, serialized.size, to_term, 0);
+}
+
+NIF(compile) {
+  ERL_NIF_TERM callback_fun_outer = argv[0];
+  LIST_PARAM(1, std::vector<mlx::core::array>, arrays);
+  ErlNifPid evaluator_pid;
+  if (!enif_get_local_pid(env, argv[2], &evaluator_pid)) {
+    return nx::nif::error(env, "Could not get evaluator pid");
+  }
+
+  ErlNifEnv *closure_env = enif_alloc_env();
+
+  auto fun = [env = closure_env, evaluator_pid, callback_fun_outer](
+                 const std::vector<mlx::core::array> &compile_args) {
+    ERL_NIF_TERM callback_fun = enif_make_copy(env, callback_fun_outer);
+    ERL_NIF_TERM tensor_list = nx::nif::make_list(env, compile_args);
+    ERL_NIF_TERM arg_list = enif_make_list1(env, tensor_list);
+
+    ERL_NIF_TERM output_list =
+        make_nif_call(env, evaluator_pid, callback_fun, arg_list);
+
+    // Convert output_list back to vector of MLX arrays
+    std::vector<mlx::core::array> output_tensors;
+    if (!nx::nif::get_list(env, output_list, output_tensors)) {
+      throw std::runtime_error("Failed to convert callback result to tensors");
+    }
+
+    enif_free_env(env);
+
+    return output_tensors;
+  };
+
+  emlx::function compiled_function_ptr = mlx::core::compile(fun);
```

Comment:
@awni I got things working via the idea we discussed yesterday! ~5x speed up vs the non-compiled version as per the benchmark in the PR description.

I'm do wonder about what XLA is doing to get such a performance boost when compiled.

Reply:
Cool, so you figured out how to pass an elixir function into C/C++?